src/lib.rs

```diff
@@ -40,12 +40,61 @@
 #![cfg_attr(feature="clippy", plugin(clippy))]
 
 extern crate libc;
+extern crate time;
 
 pub mod errors;
 pub mod backends;
 pub mod util;
 
 use errors::TraceMeError;
+use std::ops::Drop;
+use std::ptr;
+use libc::{free, c_void, getpid};
+use std::slice;
+use std::fs::File;
+use std::io::prelude::*;
+
+pub struct HWTrace {
+    buf: *const u8,
+    len: u64,
+}
+
+impl HWTrace {
+    /// Makes a new trace from a raw pointer and a size.
+    ///
+    /// The `buf` argument is assumed to have been allocated on the heap using malloc(3). `len`
+    /// must be less than or equal to the allocated size.
+    ///
+    /// Once an instance is constructed, underlying allocation and it's freeing is the
+    /// responsibility of the instance.
+    fn from_buf(buf: *const u8, len: u64) -> Self {
+        Self {buf: buf, len: len}
+    }
+
+    /// Write the trace buffer to file.
+    ///
+    /// For debugging purposes only.
+    fn to_file(&self) {
```

Comment:
This could be guarded with `#[debug]` or similar then?

Reply:
sure, didn't know about that!

Reply:
I think an earlier discussion invalidates this now. The VM author needs to be able to call this.

Reply:
Two dumb questions. Why would they need to do this? And what's to stop them writing a function to do it themselves?